server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -2543,6 +2545,39 @@ public void forceMerge(final boolean flush, int maxNumSegments, boolean onlyExpu
         }
     }
 
+    @Override
+    public boolean forceMergeIsNoOp(int maxNumSegments, boolean onlyExpungeDeletes) throws IOException {
+        if (onlyExpungeDeletes && maxNumSegments >= 0) {
+            throw new IllegalArgumentException("only_expunge_deletes and max_num_segments are mutually exclusive");
+        }
+        try (var reader = DirectoryReader.open(indexWriter)) {
+            final var segmentCommitInfos = SegmentInfos.readCommit(reader.directory(), reader.getIndexCommit().getSegmentsFileName());
```

Comment:
This looks at the `segmentInfos` in the last lucene commit, but `IndexWriter#forceMerge` does a flush and also considers segments that were generated by the flush without being yet committed.
I think this is a problem because the check might return that no force merges are required even though force merge would actually run if invoked (when there is concurrent indexing with the check).